### ISSUE TITLE
inconsistencies of DnsNameResolver.query() APIs in 4.1-beta6

### ISSUE THREAD

**Issue by dmk23** [netty/netty#4208]
I had a chance to migrate my code to the latest API (per https://github.com/netty/netty/pull/3979) and would like to share some feedback on using it in practice on a large scale.

With this mind I've implemented my own retry logic, retrying all DNS codes other than NOERROR or NXDOMAIN up to a certain number of times and reporting the final code. I've also collected stats on how many queries finish on what retry and what is the final error code. Here is an example output -

```
TOTAL: 1312129 tasks in 137.29s

NoError(0) : 1115696
FormErr(1) : 230
ServFail(2) : 4840
NXDomain(3) : 176004
timeout : 15359

Retry 0 : 1049649
Retry 1 : 189365
Retry 2 : 41463
Retry 3 : 8583
Retry 4 : 2039
Retry 5 : 21030
```

Now, the comments on usability of the current API -

1. **Timeout vs. ServFail** : From end-user perspective they are virtually the same. The difference is whether the resolver times out or returns ServFail after failing to contact the authoritative server. Yet, in terms of Netty DNS API managing this requires duplicating / splitting the error handling code that processes successful AddressedEnvelope or caught DnsNameResolverException

2. **DnsNameResolverException** : Right now the API lacks a way to return the private remoteAddress for profiling nameserver performance stats. So to build stats on nameserver performance I have to go through the absurd step of using regex to extract it from the Exception message. Generation of verbose exception messages is wholly unnecessary as well producing lots of object garbage on large scale. At the very least it could be done just-in-time by getMessage(), with each component (including remote address) available via a getter. DnsQueryContext.setFailure() should not waste CPU/RAM building exception strings

3. **Original DNS Question**: Right now it is still impossible to retrieve the original DnsQuestion object from the DnsResponse. Extracting original question from the response is not reliable since we can often see them dropped or duplicated or corrupted or missing in case of FormErr. This is even more confusing given that unlike DnsResponse, DnsNameResolverException keeps the reference to the question. Right now we have to use the ugly hack in the application by keeping a synchronized hash map of futures to questions. That should be avoided by adding an explicit reference.

4. **Retry Counters**: Right now we also have to use an extra hash map to keep track of how many times a DnsQuestion has been retried. This is clearly wasteful as well, given that it would be natural to include a retry counter directly into DnsQuestion object with get / set / increment APIs.

5. **Warning Spam** : Most warnings generated by DnsQueryContext.finish() are representing the conditions occurring normally when DNS queries are made. When operated on large scale lots of these "errors" are just FormErr messages that go away on retry. Having to do any kind of logging seems wasteful, when the application itself can handle this however it needs to.

**Proposal**: Get rid of DnsNameResolverException and merge it into DnsResponse - keep a reference to the original DnsQuestion and a user-managed retry counter. Create a "synthetic" DnsResponseCode for timeout (e.g with -1). Ensure the timeout response uses proper AddressedEnvelope

**Comment by trustin** [comment]
Thanks. Will take a look during weekends.

**Comment by dmk23** [comment]
Was this addressed in beta8 ?

**Comment by trustin** [comment]
No.  Didn't have time to evaluate the proposed change carefully yet.
I'm using it in production already, so I might have time for this
later though.

**Comment by dmk23** [comment]
Thanks. Could you at least tag this to a milestone so it does not get lost? These are serious deficiencies in the API making it harder to implement the resolver most efficiently...

**Comment by trustin** [comment]
No problem. Thanks for your patience.
